gluten-core/src/main/scala/io/glutenproject/extension/columnar/PullOutPreProject.scala

```diff
@@ -149,6 +155,49 @@ object PullOutPreProject extends Rule[SparkPlan] with PullOutProjectHelper {
         agg.child)
       newAgg.withNewChildren(Seq(preProject))
 
+    case window: WindowExec if needsPreProject(window) =>
+      val expressionMap = new mutable.HashMap[Expression, NamedExpression]()
+      // Handle orderSpec.
+      val newOrderSpec = getNewSortOrder(window.orderSpec, expressionMap)
+
+      // Handle partitionSpec.
+      val newPartitionSpec =
+        window.partitionSpec.map(replaceExpressionWithAttribute(_, expressionMap))
+
+      // Handle windowExpressions.
+      val postWindowExpressions = new ArrayBuffer[NamedExpression]()
+      val newWindowExpressions = window.windowExpression.toIndexedSeq
+        .map {
+          _.transform { case we: WindowExpression => rewriteWindowExpression(we, expressionMap) }
+        }
+        .map {
+          case alias @ Alias(_: WindowExpression, _) =>
+            postWindowExpressions += alias.toAttribute
+            alias
+          case other =>
+            // Directly use the output of WindowExpression, and move expression evaluation to
+            // post-project for computation.
+            assert(hasWindowExpression(other))
+            val we = other.collectFirst { case w: WindowExpression => w }.get
+            val rewriteWindowExpr = rewriteWindowExpression(we, expressionMap)
```

Comment:
do we need to rewrite it again? it seems we have rewritten all window expressions in previous transform line171

Reply:
Yes, no need to rewrite it again, we can just use `we`.